File: apps/ios/Domain/Tests/DomainTests/LedgerVectors.swift
import Foundation
@testable import Domain

// P1.2b: wires the real Ledger.swift port into FunctionRegistry so
// ledger.json's vectors un-skip. Registered under (domain="ledger",
// fn=<name>) to match tools/golden-vectors/vectors/ledger.json exactly --
// mirrors the Android adapter (LedgerVectors.kt) field-for-field. Called
// from VectorRunnerTests.testLedger() (XCTest).

private func asMoney(_ any: Any) -> Money {
    let dict = any as! [String: Any]
    let amount = (dict["amount"] as! NSNumber).int64Value
    let currency = dict["currency"] as! String
    return Money(amount: amount, currency: currency)
}

private func moneyToJson(_ m: Money) -> [String: Any] {
    ["amount": String(m.amount), "currency": m.currency]
}

private func asLedgerEntry(_ any: Any) -> LedgerEntry {
    let dict = any as! [String: Any]
    let toAccountId: String?
    if let raw = dict["to_account_id"], !(raw is NSNull) {
        toAccountId = raw as? String
    } else {
        toAccountId = nil
    }
    let toAmount: Int64?
    if let raw = dict["to_amount"], !(raw is NSNull) {
        toAmount = (raw as? NSNumber)?.int64Value
    } else {
        toAmount = nil
    }
    return LedgerEntry(
        type: dict["type"] as! String,
        accountId: dict["account_id"] as! String,
        amount: (dict["amount"] as! NSNumber).int64Value,
        toAccountId: toAccountId,
        toAmount: toAmount
    )
}

func registerLedgerVectors() {
    FunctionRegistry.register(domain: "ledger", fn: "signedEffectFor") { input in
        let dict = input as! [String: Any]
        let entry = asLedgerEntry(dict["entry"]!)
        let accountId = dict["accountId"] as! String
        // Stringified (like a Money amount), not a raw number -- the
        // exporter treats any raw minor-unit integer as money-shaped.
        return String(signedEffectFor(entry, accountId))
    }

    FunctionRegistry.register(domain: "ledger", fn: "deriveBalance") { input in
        let dict = input as! [String: Any]
        let accountId = dict["accountId"] as! String
        let currency = dict["currency"] as! String
        let entries = (dict["entries"] as! [Any]).map(asLedgerEntry)
        return moneyToJson(deriveBalance(accountId, currency, entries))
    }

    FunctionRegistry.register(domain: "ledger", fn: "availableBalance") { input in
        let dict = input as! [String: Any]
        return moneyToJson(try availableBalance(asMoney(dict["total"]!), asMoney(dict["blocked"]!)))
    }

    FunctionRegistry.register(domain: "ledger", fn: "aggregateNetWorth") { input in
        let dict = input as! [String: Any]
        let balances = (dict["balances"] as! [Any]).map { b -> AccountBalance in
            let bd = b as! [String: Any]
            return AccountBalance(balance: asMoney(bd["balance"]!), blocked: asMoney(bd["blocked"]!))
        }
        let base = dict["base"] as! String
        let ratesDict = dict["rates"] as! [String: Any]
        let rates: [String: Double] = ratesDict.mapValues { ($0 as! NSNumber).doubleValue }
        let includeBlocked = (dict["includeBlocked"] as! NSNumber).boolValue
        // Mirrors the TS vector's own getRate closure exactly: same
        // currency short-circuits to 1, otherwise look up `rates` and
        // default to 1 if the pair isn't present.
        let getRate: RateLookup = { from, to in from == to ? 1.0 : (rates[from] ?? 1.0) }
        return moneyToJson(try aggregateNetWorth(balances, base: base, getRate: getRate, includeBlocked: includeBlocked))
    }
}

File: apps/android/domain/src/test/kotlin/care/pocket/domain/ledger/LedgerVectors.kt
package care.pocket.domain.ledger

import care.pocket.domain.money.Money
import care.pocket.domain.vectors.FunctionRegistry
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.double
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long

// P1.2a: wires the real Ledger.kt port into FunctionRegistry so
// ledger.json's vectors un-skip. Registered under (domain="ledger",
// fn=<name>) to match tools/golden-vectors/vectors/ledger.json exactly --
// field names below (entry/accountId, account_id/to_account_id/to_amount
// on the entry itself, balances/base/rates/includeBlocked) were read
// directly off tools/golden-vectors/export.ts's ledger section, not
// guessed. Called from VectorRunnerTest.ledger() (kotlin.test).

private const val DOMAIN = "ledger"

private fun JsonElement.asMoney(): Money {
    val obj = jsonObject
    return Money(obj.getValue("amount").jsonPrimitive.long, obj.getValue("currency").jsonPrimitive.content)
}

private fun Money.toJson(): JsonElement = JsonObject(
    mapOf("amount" to JsonPrimitive(amount.toString()), "currency" to JsonPrimitive(currency))
)

private fun JsonElement.asLedgerEntry(): LedgerEntry {
    val obj = jsonObject
    return LedgerEntry(
        type = obj.getValue("type").jsonPrimitive.content,
        accountId = obj.getValue("account_id").jsonPrimitive.content,
        amount = obj.getValue("amount").jsonPrimitive.long,
        toAccountId = obj["to_account_id"]?.takeIf { it !is JsonNull }?.jsonPrimitive?.content,
        toAmount = obj["to_amount"]?.takeIf { it !is JsonNull }?.jsonPrimitive?.long,
    )
}

fun registerLedgerVectors() {
    FunctionRegistry.register(DOMAIN, "signedEffectFor") { input ->
        val obj = input.jsonObject
        val entry = obj.getValue("entry").asLedgerEntry()
        val accountId = obj.getValue("accountId").jsonPrimitive.content
        // Stringified (like a Money amount), not a raw number -- the
        // exporter treats any raw minor-unit integer as money-shaped and
        // wraps it with amt() (see export.ts's header comment).
        JsonPrimitive(signedEffectFor(entry, accountId).toString())
    }

    FunctionRegistry.register(DOMAIN, "deriveBalance") { input ->
        val obj = input.jsonObject
        val accountId = obj.getValue("accountId").jsonPrimitive.content
        val currency = obj.getValue("currency").jsonPrimitive.content
        val entries = obj.getValue("entries").jsonArray.map { it.asLedgerEntry() }
        deriveBalance(accountId, currency, entries).toJson()
    }

    FunctionRegistry.register(DOMAIN, "availableBalance") { input ->
        val obj = input.jsonObject
        availableBalance(obj.getValue("total").asMoney(), obj.getValue("blocked").asMoney()).toJson()
    }

    FunctionRegistry.register(DOMAIN, "aggregateNetWorth") { input ->
        val obj = input.jsonObject
        val balances = obj.getValue("balances").jsonArray.map { b ->
            val bo = b.jsonObject
            AccountBalance(bo.getValue("balance").asMoney(), bo.getValue("blocked").asMoney())
        }
        val base = obj.getValue("base").jsonPrimitive.content
        val rates = obj.getValue("rates").jsonObject.mapValues { it.value.jsonPrimitive.double }
        val includeBlocked = obj.getValue("includeBlocked").jsonPrimitive.boolean
        // Mirrors the TS vector's own getRate closure exactly: same
        // currency short-circuits to 1, otherwise look up `rates` and
        // default to 1 if the pair isn't present.
        val getRate: RateLookup = { from, to -> if (from == to) 1.0 else rates[from] ?: 1.0 }
        aggregateNetWorth(balances, base, getRate, includeBlocked).toJson()
    }
}
